### ISSUE TITLE
"Unsupported database file version" with 4.8 to 4.9?

### ISSUE THREAD

**Issue by freddiN** [DependencyTrack/dependency-track#3160]
### Current Behavior

Wenn migrating from 4.8 to 4.9.0 or 4.9.1, the following message is shown:

2023-10-30 14:30:41,785 ERROR [UpgradeMetaProcessor] An error occurred connecting to the database.
org.h2.jdbc.JdbcSQLNonTransientConnectionException: Unsupported database file version or invalid file header in file "/data/.dependency-track/db.mv.db" [90048-224]

<hr>
This reminded me when 4.6 was released and a migration of H2 had to be done, but I didnt find anything in the 4.9.X changelogs.

I already had this issue with 4.9.0 two weeks ago, but had put it on hold because of different priorities and I found it hard to believe to be the only one with this issue (still not 100% convinced ...)
<hr>

2023-10-30 14:30:54,445 INFO [EmbeddedJettyServer] alpine-executable-war v2.2.3 (3aff068a-a1af-4e38-85dc-3a5ee66ee121) built on: 2023-09-21T16:07:30Z
2023-10-30 14:30:55,951 INFO [Config] --------------------------------------------------------------------------------
2023-10-30 14:30:55,952 INFO [Config] OS Name:      Linux
2023-10-30 14:30:55,952 INFO [Config] OS Version:   5.10.0-13-amd64
2023-10-30 14:30:55,952 INFO [Config] OS Arch:      amd64
2023-10-30 14:30:55,952 INFO [Config] CPU Cores:    8
2023-10-30 14:30:55,954 INFO [Config] Max Memory:   9.4 GB (10,047,979,520.0 bytes)
2023-10-30 14:30:55,954 INFO [Config] Java Vendor:  Eclipse Adoptium
2023-10-30 14:30:55,955 INFO [Config] Java Version: 17.0.8.1+1
2023-10-30 14:30:55,955 INFO [Config] Java Home:    /opt/java/openjdk
2023-10-30 14:30:55,955 INFO [Config] Java Temp:    /tmp
2023-10-30 14:30:55,956 INFO [Config] User:         dtrack
2023-10-30 14:30:55,956 INFO [Config] User Home:    /data/
2023-10-30 14:30:55,956 INFO [Config] --------------------------------------------------------------------------------
2023-10-30 14:30:55,956 INFO [Config] Initializing Configuration
2023-10-30 14:30:55,957 INFO [Config] System property alpine.application.properties not specified
2023-10-30 14:30:55,957 INFO [Config] Loading application.properties from classpath
2023-10-30 14:30:55,960 INFO [Config] --------------------------------------------------------------------------------
2023-10-30 14:30:55,960 INFO [Config] Application:  Dependency-Track
2023-10-30 14:30:55,960 INFO [Config] Version:      4.9.1
2023-10-30 14:30:55,960 INFO [Config] Built-on:     2023-10-30T12:04:30Z
2023-10-30 14:30:55,960 INFO [Config] --------------------------------------------------------------------------------
2023-10-30 14:30:55,960 INFO [Config] Framework:    Alpine
2023-10-30 14:30:55,961 INFO [Config] Version :     2.2.3
2023-10-30 14:30:55,961 INFO [Config] Built-on:     2023-09-21T16:07:30Z
2023-10-30 14:30:55,961 INFO [Config] --------------------------------------------------------------------------------

<hr>
Full stack trace:

2023-10-30 14:30:41,785 ERROR [UpgradeMetaProcessor] An error occurred connecting to the database.
org.h2.jdbc.JdbcSQLNonTransientConnectionException: Unsupported database file version or invalid file header in file "/data/.dependency-track/db.mv.db" [90048-224]
        at org.h2.message.DbException.getJdbcSQLException(DbException.java:690)
        at org.h2.message.DbException.getJdbcSQLException(DbException.java:489)
        at org.h2.message.DbException.get(DbException.java:212)
        at org.h2.mvstore.db.Store.convertMVStoreException(Store.java:158)
        at org.h2.mvstore.db.Store.<init>(Store.java:142)
        at org.h2.engine.Database.<init>(Database.java:326)
        at org.h2.engine.Engine.openSession(Engine.java:92)
        at org.h2.engine.Engine.openSession(Engine.java:222)
        at org.h2.engine.Engine.createSession(Engine.java:201)
        at org.h2.engine.SessionRemote.connectEmbeddedOrServer(SessionRemote.java:343)
        at org.h2.jdbc.JdbcConnection.<init>(JdbcConnection.java:125)
        at org.h2.Driver.connect(Driver.java:59)
        at java.sql/java.sql.DriverManager.getConnection(Unknown Source)
        at java.sql/java.sql.DriverManager.getConnection(Unknown Source)
        at alpine.server.upgrade.UpgradeMetaProcessor.createConnection(UpgradeMetaProcessor.java:187)
        at alpine.server.upgrade.UpgradeMetaProcessor.<init>(UpgradeMetaProcessor.java:63)
        at org.dependencytrack.upgrade.UpgradeInitializer.contextInitialized(UpgradeInitializer.java:56)
        at org.eclipse.jetty.server.handler.ContextHandler.callContextInitialized(ContextHandler.java:1050)
        at org.eclipse.jetty.servlet.ServletContextHandler.callContextInitialized(ServletContextHandler.java:624)
        at org.eclipse.jetty.server.handler.ContextHandler.contextInitialized(ContextHandler.java:985)
        at org.eclipse.jetty.servlet.ServletHandler.initialize(ServletHandler.java:740)
        at org.eclipse.jetty.servlet.ServletContextHandler.startContext(ServletContextHandler.java:392)
        at org.eclipse.jetty.webapp.WebAppContext.startContext(WebAppContext.java:1304)
        at org.eclipse.jetty.server.handler.ContextHandler.doStart(ContextHandler.java:902)
        at org.eclipse.jetty.servlet.ServletContextHandler.doStart(ServletContextHandler.java:306)
        at org.eclipse.jetty.webapp.WebAppContext.doStart(WebAppContext.java:532)
        at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:93)
        at org.eclipse.jetty.util.component.ContainerLifeCycle.start(ContainerLifeCycle.java:171)
        at org.eclipse.jetty.server.Server.start(Server.java:470)
        at org.eclipse.jetty.util.component.ContainerLifeCycle.doStart(ContainerLifeCycle.java:114)
        at org.eclipse.jetty.server.handler.AbstractHandler.doStart(AbstractHandler.java:89)
        at org.eclipse.jetty.server.Server.doStart(Server.java:415)
        at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:93)
        at alpine.embedded.EmbeddedJettyServer.main(EmbeddedJettyServer.java:100)
Caused by: org.h2.mvstore.MVStoreException: The write format 2 is smaller than the supported format 3 [2.2.224/5]
Caused by: org.h2.jdbc.JdbcSQLNonTransientConnectionException: Unsupported database file version or invalid file header in file "/data/.dependency-track/db.mv.db" [90048-224]
        at org.h2.message.DbException.getJdbcSQLException(DbException.java:690)
        at org.h2.message.DbException.getJdbcSQLException(DbException.java:489)
        at org.h2.message.DbException.get(DbException.java:212)
        at org.h2.mvstore.db.Store.convertMVStoreException(Store.java:158)
        at org.h2.mvstore.db.Store.<init>(Store.java:142)
        at org.h2.engine.Database.<init>(Database.java:326)
        at org.h2.engine.Engine.openSession(Engine.java:92)
        at org.h2.engine.Engine.openSession(Engine.java:222)
        at org.h2.engine.Engine.createSession(Engine.java:201)
        at org.h2.engine.SessionRemote.connectEmbeddedOrServer(SessionRemote.java:343)
        at org.h2.jdbc.JdbcConnection.<init>(JdbcConnection.java:125)
        at org.h2.Driver.connect(Driver.java:59)
        at java.sql/java.sql.DriverManager.getConnection(Unknown Source)
        at java.sql/java.sql.DriverManager.getConnection(Unknown Source)
        at alpine.server.upgrade.UpgradeMetaProcessor.createConnection(UpgradeMetaProcessor.java:187)
        ... 19 common frames omitted
Caused by: org.h2.mvstore.MVStoreException: The write format 2 is smaller than the supported format 3 [2.2.224/5]
        at org.h2.mvstore.DataUtils.newMVStoreException(DataUtils.java:996)
        at org.h2.mvstore.FileStore.getUnsupportedWriteFormatException(FileStore.java:943)
        at org.h2.mvstore.FileStore.processCommonHeaderAttributes(FileStore.java:547)
        at org.h2.mvstore.RandomAccessStore.readStoreHeader(RandomAccessStore.java:227)
        at org.h2.mvstore.FileStore.start(FileStore.java:916)
        at org.h2.mvstore.MVStore.<init>(MVStore.java:289)
        at org.h2.mvstore.MVStore$Builder.open(MVStore.java:2035)
        at org.h2.mvstore.db.Store.<init>(Store.java:133)
        ... 29 common frames omitted

### Steps to Reproduce

1. 4.8 docker setup with h2 db
2. switct the image to 4.9.0 or 4.9.1

### Expected Behavior

Automatic db migration

### Dependency-Track Version

4.9.1

### Dependency-Track Distribution

Container Image

### Database Server

H2

### Browser

Google Chrome

### Checklist

- [X] I have read and understand the [contributing guidelines](https://github.com/DependencyTrack/dependency-track/blob/master/CONTRIBUTING.md#filing-issues)
- [X] I have checked the [existing issues](https://github.com/DependencyTrack/dependency-track/issues) for whether this defect was already reported

**Comment by ataraxus** [comment]
Same issue here: 
`dtrack-apiserver_1  | Caused by: org.h2.jdbc.JdbcSQLNonTransientConnectionException: Unsupported database file version or invalid file header in file "/data/.dependency-track/db.mv.db" [90048-224]`

**Comment by nscuro** [comment]
Between 4.8.x and 4.9.x, the `h2` dependency was bumped from `2.1.214` to `2.2.220`. I couldn't find anything indicating a breaking change in the database format in the changelogs: https://www.h2database.com/html/changelog.html

That said, the embedded H2 database was never intended for production use, and the documentation clearly states that. You can likely do something similar to [what was done for the H2 v1 -> v2 migration](https://docs.dependencytrack.org/getting-started/database-support/#migrating-to-h2-v2).

**Comment by freddiN** [comment]
I found this interesting project to migrate between h2 versions: https://github.com/manticore-projects/H2MigrationTool
I will now go ahead and try to migrate this way. If I have any interesting findings, Ill post them here.

For future references I would note that a change like this should be mentioned in the changelog like for 4.6, but then again I m pretty sure this was an oversight anyway, so not sure what can be learned from it ...

**Comment by freddiN** [comment]
I could not get the tool to work, so I went with the steps described here:
https://docs.dependencytrack.org/getting-started/database-support/#migrating-to-h2-v2

Make sure to use the 2.1.214 jar for creating the backup and 2.2.220 for importing the backup. Everything else works exactly as described in the guide above, maybe with one addition:
If you see "The database is read only" messages after the restart, the created db has the wrong access rights.
I ran the migration as root but docker uses uid 1000, so a "chown 1000:1000 db.mv.db" was in order before the dependency track restart.
